pkg/destroy/aws/aws.go

```diff
@@ -77,1510 +63,1205 @@ func (o *ClusterUninstaller) validate() error {
-	s, err := session.NewSession(awsConfig)
 	if err != nil {
-		return nil, err
+		return arns, err
 	}
-
-	return s, nil
+	return arns, err2
 }
 
-func createEC2Filters(filters Filter) []*ec2.Filter {
-	Filter := []*ec2.Filter{}
-	for key, val := range filters {
-		Filter = append(Filter, &ec2.Filter{
-			Name:   aws.String(fmt.Sprintf("tag:%s", key)),
-			Values: []*string{aws.String(val)},
-		})
-	}
-
-	return Filter
+type iamUserSearch struct {
+	client    *iam.IAM
+	filters   []Filter
+	logger    logrus.FieldLogger
+	unmatched map[string]struct{}
 }
 
-// tagsToMap takes various types of AWS-object tags and returns a map-representation
-func tagsToMap(tags interface{}) (map[string]string, error) {
-	x := map[string]string{}
-
-	switch v := tags.(type) {
-	case []*autoscaling.TagDescription:
-		for _, tag := range v {
-			x[*tag.Key] = *tag.Value
-		}
-	case *elb.TagDescription:
-		for _, tag := range v.Tags {
-			x[*tag.Key] = *tag.Value
-		}
-	case []*s3.Tag:
-		for _, tag := range v {
-			x[*tag.Key] = *tag.Value
-		}
-	case []*route53.Tag:
-		for _, tag := range v {
-			x[*tag.Key] = *tag.Value
-		}
-	case []*iam.Tag:
-		for _, tag := range v {
-			x[*tag.Key] = *tag.Value
-		}
-	default:
-		return x, errors.Errorf("unable to convert type: %v", v)
+func (search *iamUserSearch) arns() ([]string, error) {
+	if search.unmatched == nil {
+		search.unmatched = map[string]struct{}{}
 	}
 
-	return x, nil
-}
+	arns := []string{}
+	var err error
+	err2 := search.client.ListUsersPages(
+		&iam.ListUsersInput{},
+		func(results *iam.ListUsersOutput, lastPage bool) bool {
+			for _, user := range results.Users {
+				if _, ok := search.unmatched[*user.Arn]; ok {
+					continue
+				}
+
+				// Unfortunately user.Tags is empty from ListUsers, so we need to query each one
+				var response *iam.GetUserOutput
+				response, err = search.client.GetUser(&iam.GetUserInput{UserName: aws.String(*user.UserName)})
+				if err != nil {
+					if err.(awserr.Error).Code() == iam.ErrCodeNoSuchEntityException {
+						search.unmatched[*user.Arn] = exists
+					} else {
+						err = errors.Wrapf(err, "get tags for %s", *user.Arn)
+						search.logger.Info(err)
+					}
+				} else {
+					user = response.User
+					tags := make(map[string]string, len(user.Tags))
+					for _, tag := range user.Tags {
+						tags[*tag.Key] = *tag.Value
+					}
+					if tagMatch(search.filters, tags) {
+						arns = append(arns, *user.Arn)
+					} else {
+						search.unmatched[*user.Arn] = exists
+					}
+				}
+			}
 
-// filterLBsByVPC will find all the load balancers in the provided list that are under the provided VPC
-func filterLBsByVPC(lbs []*elb.LoadBalancerDescription, vpc *ec2.Vpc, logger logrus.FieldLogger) []*elb.LoadBalancerDescription {
-	filteredLBs := []*elb.LoadBalancerDescription{}
+			return !lastPage
+		},
+	)
 
-	for _, lb := range lbs {
-		if *lb.VPCId == *vpc.VpcId {
-			filteredLBs = append(filteredLBs, lb)
-		}
+	if err != nil {
```

Comment:
Same as above, should we fail on last failed `search.client.GetUser(&iam.GetUserInput{UserName: aws.String(*user.UserName)})`

Reply:
Discussion [here][1].

[1]: https://github.com/openshift/installer/pull/1039#discussion_r249564271